internal/release/github.go

```diff
@@ -0,0 +1,171 @@
+		return nil, err
+	}
+	return release, nil
+}
+
+// UploadReleaseAsset uploads the specified file to the release.
+func (c *Client) UploadReleaseAsset(ctx context.Context, owner GithubOwner, repo GithubRepo, releaseID int64, filename string) error {
+	f, err := os.Open(filename)
+	if err != nil {
+		return err
+	}
+	_, _, err = c.GitHub.Repositories.UploadReleaseAsset(ctx, string(owner), string(repo), releaseID, &github.UploadOptions{
+		Name: filepath.Base(filename),
+	}, f)
+	return err
+}
+
+// GetLatestRelease returns information about the latest GitHub release for the given org and repo (i.e. 'bufbuild', 'plugins').
+// If no release is found, returns ErrReleaseNotFound.
+func (c *Client) GetLatestRelease(ctx context.Context, owner GithubOwner, repo GithubRepo) (*github.RepositoryRelease, error) {
+	repositoryRelease, resp, err := c.GitHub.Repositories.GetLatestRelease(ctx, string(owner), string(repo))
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			// no latest release
+			return nil, ErrNotFound
+		}
+		return nil, err
+	}
+	return repositoryRelease, nil
+}
+
+// GetReleaseByTag returns information about a given release (by tag name).
+func (c *Client) GetReleaseByTag(ctx context.Context, owner GithubOwner, repo GithubRepo, tag string) (*github.RepositoryRelease, error) {
+	repositoryRelease, resp, err := c.GitHub.Repositories.GetReleaseByTag(ctx, string(owner), string(repo), tag)
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			// no latest release
+			return nil, ErrNotFound
+		}
+		return nil, err
+	}
+	return repositoryRelease, nil
+}
+
+func getOrgRepoFromReleaseURL(url string) (string, string, error) {
+	_, orgRepo, found := strings.Cut(url, "/repos/")
+	if !found {
+		return "", "", fmt.Errorf("unsupported release URL format - no /repos/ found: %q", url)
+	}
+	orgRepo, _, found = strings.Cut(orgRepo, "/releases/")
+	if !found {
+		return "", "", fmt.Errorf("unsupported release URL format - no /releases/ found: %q", url)
+	}
+	org, repo, found := strings.Cut(orgRepo, "/")
+	if !found {
+		return "", "", fmt.Errorf("unsupported release URL format no org/repo found: %q", url)
+	}
+	return org, repo, nil
+}
+
+// LoadPluginReleases loads the plugin-releases.json file from the specified GitHub release.
+// It will additionally verify the minisign signature of the release if passed a valid minisign.PublicKey.
+func (c *Client) LoadPluginReleases(ctx context.Context, release *github.RepositoryRelease, publicKey minisign.PublicKey) (*PluginReleases, error) {
+	releasesJSONBytes, err := c.downloadAsset(ctx, release, PluginReleasesFile)
+	if err != nil {
+		return nil, err
+	}
+	releasesJSONMinisigBytes, err := c.downloadAsset(ctx, release, PluginReleasesSignatureFile)
+	if err != nil && !errors.Is(err, ErrNotFound) {
+		return nil, err
+	}
+	if !publicKey.Equal(minisign.PublicKey{}) && !minisign.Verify(publicKey, releasesJSONBytes, releasesJSONMinisigBytes) {
```

Comment:
Now if we're passed a valid `minisign.PublicKey`, we verify it first before parsing it. This means that as long as keys aren't changing, we'll always verify contents before creating a new release based on this.

Reply:
Does this mean if we rotate the key, this will break because of a mismatch with previous release?

Reply:
Yep - if we rotate the key we might need to unblock releases by adding the new key in addition to the old key, and supporting both. Once we've released w/ the new key, we could remove the old key.